lib/DxilValidation/DxilContainerValidation.cpp

```diff
@@ -73,6 +76,667 @@ struct DiagRestore {
 static void emitDxilDiag(LLVMContext &Ctx, const char *str) {
   hlsl::dxilutil::EmitErrorOnContext(Ctx, str);
 }
+
+struct SimpleViewIDState {
+  unsigned NumInputSigScalars = 0;
+  unsigned NumOutputSigScalars[DXIL::kNumOutputStreams] = {0, 0, 0, 0};
+  unsigned NumPCOrPrimSigScalars = 0;
+
+  ArrayRef<uint32_t> InputToOutputTable[DXIL::kNumOutputStreams];
+  ArrayRef<uint32_t> InputToPCOutputTable;
+  ArrayRef<uint32_t> PCInputToOutputTable;
+  ArrayRef<uint32_t> ViewIDOutputMask[DXIL::kNumOutputStreams];
+  ArrayRef<uint32_t> ViewIDPCOutputMask;
+  bool IsValid = true;
+  SimpleViewIDState(std::vector<uint32_t> &Data, PSVShaderKind,
+                    bool UsesViewID);
+};
+static unsigned RoundUpToUINT(unsigned x) { return (x + 31) / 32; }
+SimpleViewIDState::SimpleViewIDState(std::vector<uint32_t> &Data,
+                                     PSVShaderKind SK, bool UsesViewID) {
+  if (Data.empty())
+    return;
+
+  unsigned Offset = 0;
+  // #Inputs.
+  NumInputSigScalars = Data[Offset++];
+  unsigned NumStreams =
+      SK == PSVShaderKind::Geometry ? DXIL::kNumOutputStreams : 1;
+  for (unsigned i = 0; i < NumStreams; i++) {
+    // #Outputs for stream StreamId.
+    unsigned OutputScalars = Data[Offset++];
+    NumOutputSigScalars[i] = OutputScalars;
+    unsigned MaskDwords = RoundUpToUINT(OutputScalars);
+    if (UsesViewID) {
+      ViewIDOutputMask[i] = {Data.data() + Offset, MaskDwords};
```

Comment:
Can you do this with `ArrayRef::operator[]` instead? That gives us bounds checking.

Reply:
Something like this?
```
ViewIDOutputMask[i] = {&(Data[Offset]), MaskDwords};
```